File: tests/benchmark/common/benchmark_framework.cpp
#include "benchmark/common/benchmark_framework.h"
#include <iostream>
#include <iomanip>
#include <algorithm>
#include <sstream>
#include <getopt.h>
#include "origin/cuda/cuda.h"

namespace origin {
namespace benchmark {

std::string BenchmarkFramework::device_type_to_string(DeviceType device_type)
{
    return (device_type == DeviceType::kCPU) ? "cpu" : "cuda";
}

std::vector<std::vector<Shape>> BenchmarkFramework::get_default_shapes() const
{
    // 对于单shape算子，提供默认实现
    if (get_required_shapes_count() == 1)
    {
        return {
            {Shape({1, 1})},
            {Shape({10, 10})},
            {Shape({100, 100})},
            {Shape({1000, 1000})},
            {Shape({10000, 10000})},
        };
    }
    // 对于多shape算子，子类必须重写此方法
    THROW_RUNTIME_ERROR("get_default_shapes() must be overridden for operators requiring {} shapes", get_required_shapes_count());
}

void BenchmarkFramework::usage(const char* program_name) const
{
    loga("Usage: {} [OPTIONS]", program_name);
    loga("Options:");
    loga("  -d, --device DEVICE       Device type: cpu or cuda (can be specified multiple times)");
    loga("                            If not specified, tests all available devices");
    loga("  -s, --shape SHAPE         Tensor shape, e.g., \"100,100\" or \"1000,1000\"");
    loga("                            For operators requiring multiple shapes, use ':' to separate");
    loga("                            e.g., \"100,200:200,50\" for matmul");
    loga("                            (can be specified multiple times)");
    loga("                            If not specified, uses default shapes");
    std::string additional_help = get_additional_help();
    if (!additional_help.empty())
    {
        loga("{}", additional_help);
    }
    loga("  -w, --warmup ITERATIONS   Number of warmup iterations (default: 5)");
    loga("  -r, --repeat ITERATIONS   Number of repeat iterations (default: 100)");
    loga("  -h, --help                Show this help message");
}

bool BenchmarkFramework::parse_arguments(int argc, char* argv[],
                                         std::vector<std::vector<Shape>>& shapes_list,
                                         std::vector<DeviceType>& devices,
                                         int& warmup_cnt,
                                         int& repeat_cnt) const
{
    bool use_default_shapes = true;
    bool use_default_devices = true;
    warmup_cnt = 5;
    repeat_cnt = 100;
    size_t required_shapes_count = get_required_shapes_count();

    // 解析命令行参数
    static struct option long_options[] = {
        {"device", required_argument, 0, 'd'},
        {"shape", required_argument, 0, 's'},
        {"warmup", required_argument, 0, 'w'},
        {"repeat", required_argument, 0, 'r'},
        {"help", no_argument, 0, 'h'},
        {0, 0, 0, 0}
    };

    int option_index = 0;
    int c;

    while ((c = getopt_long(argc, argv, "d:s:w:r:h", long_options, &option_index)) != -1)
    {
        switch (c)
        {
            case 'd':
            {
                std::string device_str = optarg;
                if (device_str == "cpu")
                {
                    if (use_default_devices)
                    {
                        devices.clear();
                        use_default_devices = false;
                    }
                    devices.push_back(DeviceType::kCPU);
                }
                else if (device_str == "cuda")
                {
                    if (use_default_devices)
                    {
                        devices.clear();
                        use_default_devices = false;
                    }
                    devices.push_back(DeviceType::kCUDA);
                }
                else
                {
                    loge("Error: Invalid device '{}'. Use 'cpu' or 'cuda'.", device_str);
                    return false;
                }
                break;
            }
            case 's':
            {
                std::string shape_str = optarg;
                std::vector<Shape> shapes;
                
                // 检查是否包含冒号（多个shape）
                if (shape_str.find(':') != std::string::npos)
                {
                    // 解析多个shape
                    shapes = parse_multiple_shapes_string(shape_str);
                    if (shapes.empty())
                    {
                        loge("Error: Invalid shapes '{}'. Expected format: 'dim1,dim2,...:dim1,dim2,...'", shape_str);
                        return false;
                    }
                }
                else
                {
                    // 解析单个shape
                    Shape shape = parse_shape_string(shape_str);
                    if (shape.ndims() == 0)
                    {
                        loge("Error: Invalid shape '{}'. Expected format: 'dim1,dim2,...'", shape_str);
                        return false;
                    }
                    shapes.push_back(shape);
                }
                
                // 验证shape数量
                if (shapes.size() != required_shapes_count)
                {
                    loge("Error: Operator requires {} shape(s), but got {} shape(s) in '{}'",
                         required_shapes_count, shapes.size(), shape_str);
                    return false;
                }
                
                // 验证形状（子类可以重写此方法）
                try
                {
                    validate_shapes(shapes);
                }
                catch (const std::exception& e)
                {
                    loge("Error: Invalid shapes '{}': {}", shape_str, e.what());
                    return false;
                }
                
                if (use_default_shapes)
                {
                    shapes_list.clear();
                    use_default_shapes = false;
                }
                shapes_list.push_back(shapes);
                break;
            }
            case 'w':
            {
                try
                {
                    warmup_cnt = std::stoi(optarg);
                    if (warmup_cnt < 0)
                    {
                        loge("Error: Warmup count must be non-negative");
                        return false;
                    }
                }
                catch (const std::exception&)
                {
                    loge("Error: Invalid warmup count '{}'", optarg);
                    return false;
                }
                break;
            }
            case 'r':
            {
                try
                {
                    repeat_cnt = std::stoi(optarg);
                    if (repeat_cnt <= 0)
                    {
                        loge("Error: Repeat count must be positive");
                        return false;
                    }
                }
                catch (const std::exception&)
                {
                    loge("Error: Invalid repeat count '{}'", optarg);
                    return false;
                }
                break;
            }
            case 'h':
                usage(argv[0]);
                return false;  // 返回false表示需要退出
            case '?':
                loga("Use -h or --help for usage information");
                return false;
            default:
                break;
        }
    }

    // 如果使用默认shapes，则使用默认列表
    if (use_default_shapes)
    {
        shapes_list = get_default_shapes();
    }

    // 处理设备列表
    process_devices(devices, use_default_devices);

    if (shapes_list.empty())
    {
        loge("Error: No shapes specified");
        return false;
    }

    if (devices.empty())
    {
        loge("Error: No devices specified");
        return false;
    }

    return true;
}

void BenchmarkFramework::process_devices(std::vector<DeviceType>& devices, bool use_default_devices) const
{
    if (use_default_devices)
    {
        devices.push_back(DeviceType::kCPU);
        if (cuda::is_available())
        {
            devices.push_back(DeviceType::kCUDA);
        }
    }
    else
    {
        // 检查指定的CUDA设备是否可用
        bool has_cuda = false;
        for (auto device : devices)
        {
            if (device == DeviceType::kCUDA)
            {
                has_cuda = true;
                break;
            }
        }

        if (has_cuda)
        {
            if (!cuda::is_available())
            {
                logw("Warning: CUDA is not available, skipping CUDA tests");
                devices.erase(std::remove(devices.begin(), devices.end(), DeviceType::kCUDA), devices.end());
            }
        }
    }
}

int BenchmarkFramework::run(int argc, char* argv[])
{
    std::vector<std::vector<Shape>> shapes_list;
    std::vector<DeviceType> devices;
    int warmup_cnt;
    int repeat_cnt;

    // 解析命令行参数
    if (!parse_arguments(argc, argv, shapes_list, devices, warmup_cnt, repeat_cnt))
    {
        // parse_arguments 返回 false 表示需要退出（如显示帮助）
        return 0;
    }

    std::vector<DataType> dtypes = {DataType::kFloat32};

    // 先收集所有结果和字符串，用于计算列宽
    struct ResultRow {
        std::string shape_str;
        std::string repeat_str;
        std::string device_str;
        std::string dtype_str;
        std::string time_str;
        bool valid;
    };
    std::vector<ResultRow> results;

    // 运行测试并收集结果
    for (const auto& shapes : shapes_list)
    {
        for (const auto& dtype : dtypes)
        {
            for (const auto& device_type : devices)
            {
                ResultRow row;
                row.valid = false;

                // 构造shape字符串
                std::string shape_str;
                for (size_t i = 0; i < shapes.size(); ++i)
                {
                    if (i > 0)
                    {
                        shape_str += ":";
                    }
                    shape_str += shapes[i].to_string();
                }
                row.shape_str = shape_str;

                try
                {
                    BenchmarkConfig config{
                        shapes,
                        dtype,
                        Device(device_type, 0),
                        warmup_cnt,
                        repeat_cnt
                    };

                    double avg_time_us = run_benchmark(config);

                    row.repeat_str = std::to_string(repeat_cnt);
                    row.device_str = device_type_to_string(device_type);
                    row.dtype_str = origin::dtype_to_string(dtype);
                    
                    std::ostringstream oss;
                    oss << std::fixed << std::setprecision(4) << avg_time_us;
                    row.time_str = oss.str();
                    row.valid = true;
                }
                catch (const std::exception& e)
                {
                    loge("Error testing {} {} {}: {}",
                         shape_str,
                         device_type_to_string(device_type),
                         origin::dtype_to_string(dtype),
                         e.what());
                    // row.valid 保持为 false，跳过这一行
                }

                if (row.valid)
                {
                    results.push_back(row);
                }
            }
        }
    }

    // 计算每列的最大宽度
    size_t max_shape_width = 5;  // "shape" 长度
    size_t max_repeat_width = 6; // "repeat" 长度
    size_t max_device_width = 6; // "device" 长度
    size_t max_dtype_width = 5;  // "dtype" 长度
    size_t max_time_width = 17;  // "origindl_time_us" 长度

    for (const auto& row : results)
    {
        max_shape_width = std::max(max_shape_width, row.shape_str.length());
        max_repeat_width = std::max(max_repeat_width, row.repeat_str.length());
        max_device_width = std::max(max_device_width, row.device_str.length());
        max_dtype_width = std::max(max_dtype_width, row.dtype_str.length());
        max_time_width = std::max(max_time_width, row.time_str.length());
    }

    // 输出表头（制表符分隔，方便Python解析，列宽增加3个字符以增加间距）
    // 注意：制表符会在每个tab stop对齐，为了确保间距足够，我们在列宽基础上增加3
    const int extra_spacing = 3;
    std::cout << std::left
              << std::setw(static_cast<int>(max_shape_width + extra_spacing)) << "shape" << "\t"
              << std::setw(static_cast<int>(max_repeat_width + extra_spacing)) << "repeat" << "\t"
              << std::setw(static_cast<int>(max_device_width + extra_spacing)) << "device" << "\t"
              << std::setw(static_cast<int>(max_dtype_width + extra_spacing)) << "dtype" << "\t"
              << std::setw(static_cast<int>(max_time_width + extra_spacing)) << "origindl_time_us"
              << std::endl;

    // 输出结果（使用制表符分隔，方便Python解析）
    for (const auto& row : results)
    {
        std::cout << std::left
                  << std::setw(static_cast<int>(max_shape_width + extra_spacing)) << row.shape_str << "\t"
                  << std::setw(static_cast<int>(max_repeat_width + extra_spacing)) << row.repeat_str << "\t"
                  << std::setw(static_cast<int>(max_device_width + extra_spacing)) << row.device_str << "\t"
                  << std::setw(static_cast<int>(max_dtype_width + extra_spacing)) << row.dtype_str << "\t"
                  << std::setw(static_cast<int>(max_time_width + extra_spacing)) << row.time_str
                  << std::endl;
    }

    return 0;
}

}  // namespace benchmark
}  // namespace origin

File: tests/benchmark/common/benchmark_framework.py
"""
Benchmark framework for PyTorch performance testing
学习C++ benchmark框架的设计，提供统一的Python性能测试接口
"""

import sys
import torch
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import List, Dict, Optional, Tuple
from parser_utils import parse_shape_string, parse_multiple_shapes_string
from timer import Timer


@dataclass
class BenchmarkConfig:
    """基准测试配置"""
    shapes: List[Tuple[int, ...]]  # 输入的shape列表（单shape或多shape）
    dtype: str = 'float32'          # 数据类型
    device: str = 'cpu'             # 设备（'cpu'或'cuda:0'等）
    warmup_cnt: int = 5             # 预热次数
    repeat_cnt: int = 100           # 重复次数
    
    def __post_init__(self):
        """确保warmup_cnt和repeat_cnt有默认值"""
        if self.warmup_cnt is None:
            self.warmup_cnt = 5
        if self.repeat_cnt is None:
            self.repeat_cnt = 100


class BenchmarkFramework(ABC):
    """基准测试框架基类
    
    提供命令行参数解析、测试循环等通用功能
    子类需要实现 run_benchmark() 方法来执行具体的PyTorch基准测试
    """
    
    @abstractmethod
    def get_required_shapes_count(self) -> int:
        """获取所需的shape数量（纯虚函数，由子类实现）
        
        Returns:
            所需的shape数量，例如add算子返回1，matmul算子返回2
        """
        pass
    
    @abstractmethod
    def run_benchmark(self, config: BenchmarkConfig) -> float:
        """运行基准测试（纯虚函数，由子类实现）
        
        Args:
            config: 基准测试配置
            
        Returns:
            平均执行时间（微秒）
        """
        pass
    
    def get_default_shapes(self) -> List[List[Tuple[int, ...]]]:
        """获取默认测试形状
        
        Returns:
            默认形状列表，每个元素是一个shape组合
            对于单shape算子，每个元素是单个shape的列表
            默认实现：对于单shape算子，返回 [(1,1)], [(10,10)], [(100,100)], [(1000,1000)], [(10000,10000)]
            多shape算子需要重写此方法
        """
        if self.get_required_shapes_count() == 1:
            return [
                [(1, 1)],
                [(10, 10)],
                [(100, 100)],
                [(1000, 1000)],
                [(10000, 10000)],
            ]
        else:
            raise NotImplementedError(
                f"get_default_shapes() must be overridden for operators requiring "
                f"{self.get_required_shapes_count()} shapes"
            )
    
    def validate_shapes(self, shapes: List[Tuple[int, ...]]) -> None:
        """验证形状是否有效（可选，子类可重写）
        
        Args:
            shapes: 要验证的形状列表
            
        Raises:
            ValueError: 如果形状无效
        """
        pass
    
    def get_operator_name(self) -> str:
        """获取算子名称（用于帮助信息）
        
        Returns:
            算子名称字符串
        """
        return self.__class__.__name__.replace('Benchmark', '')
    
    def _parse_device(self, device_str: str) -> torch.device:
        """解析设备字符串为torch.device对象
        
        Args:
            device_str: 设备字符串，例如 'cpu' 或 'cuda:0'
            
        Returns:
            torch.device对象
        """
        if device_str == 'cpu':
            return torch.device('cpu')
        elif device_str.startswith('cuda:'):
            index = int(device_str.split(':')[1])
            return torch.device(f'cuda:{index}')
        else:
            raise ValueError(f"Invalid device string: {device_str}")
    
    def _get_available_devices(self, device_filter: Optional[str] = None) -> List[str]:
        """获取可用设备列表
        
        Args:
            device_filter: 设备过滤，'cpu'或'cuda'，None表示所有设备
            
        Returns:
            可用设备列表，例如 ['cpu', 'cuda:0']
        """
        devices = ['cpu']
        
        if torch.cuda.is_available():
            device_count = torch.cuda.device_count()
            for i in range(device_count):
                devices.append(f'cuda:{i}')
        
        # 应用过滤
        if device_filter == 'cpu':
            return ['cpu']
        elif device_filter == 'cuda':
            return [d for d in devices if d.startswith('cuda')]
        elif device_filter:
            # 如果指定了具体设备，检查是否可用
            if device_filter in devices:
                return [device_filter]
            else:
                raise ValueError(f"Device '{device_filter}' is not available")
        
        return devices
    
    def _format_shape_string(self, shapes: List[Tuple[int, ...]]) -> str:
        """格式化shape为字符串，例如 [(100, 100)] -> "{100,100}" 或 [(100, 200), (200, 50)] -> "{100,200}:{200,50}"
        
        Args:
            shapes: shape列表
            
        Returns:
            格式化的shape字符串
        """
        shape_strs = []
        for shape in shapes:
            shape_str = '{' + ','.join(str(d) for d in shape) + '}'
            shape_strs.append(shape_str)
        return ':'.join(shape_strs)
    
    def _parse_shapes_from_string(self, shape_filter: str) -> List[Tuple[int, ...]]:
        """从字符串解析shapes
        
        Args:
            shape_filter: shape字符串，例如 '1000,1000' 或 '100,200:200,50'
            
        Returns:
            解析后的shape列表
        """
        required_count = self.get_required_shapes_count()
        
        # 检查是否包含冒号（多个shape）
        if ':' in shape_filter:
            shapes = parse_multiple_shapes_string(shape_filter)
        else:
            # 单个shape
            dims = parse_shape_string(shape_filter)
            if not dims:
                raise ValueError(f"Invalid shape string: {shape_filter}")
            shapes = [tuple(dims)]
        
        # 验证shape数量
        if len(shapes) != required_count:
            raise ValueError(
                f"Operator requires {required_count} shape(s), but got {len(shapes)} shape(s) in '{shape_filter}'"
            )
        
        # 验证形状（子类可以重写此方法）
        self.validate_shapes(shapes)
        
        return shapes
    
    def run(self,
            device_filter: Optional[str] = None,
            shape_filter: Optional[str] = None,
            warmup_cnt: int = 5,
            repeat_cnt: int = 100,
            verbose: bool = False) -> List[Dict]:
        """运行基准测试主函数
        
        解析参数，运行测试循环，返回结果列表
        与C++框架的run()方法功能类似
        
        Args:
            device_filter: 设备过滤（'cpu'或'cuda'），None表示所有设备
            shape_filter: shape过滤（例如'1000,1000'），None表示使用默认shapes
            warmup_cnt: 预热次数
            repeat_cnt: 重复次数
            verbose: 是否输出详细信息
            
        Returns:
            PyTorch测试结果列表，每个元素为:
            {'shape': str, 'device': str, 'dtype': str, 'time_us': float}
        """
        # 获取可用设备
        devices = self._get_available_devices(device_filter)
        
        # 获取测试shapes
        if shape_filter:
            shapes_list = [self._parse_shapes_from_string(shape_filter)]
        else:
            shapes_list = self.get_default_shapes()
        
        # 测试数据类型
        dtypes = ['float32']
        
        # 运行测试并收集结果
        results = []
        
        for shapes in shapes_list:
            for dtype in dtypes:
                for device_str in devices:
                    try:
                        config = BenchmarkConfig(
                            shapes=shapes,
                            dtype=dtype,
                            device=device_str,
                            warmup_cnt=warmup_cnt,
                            repeat_cnt=repeat_cnt
                        )
                        
                        avg_time_us = self.run_benchmark(config)
                        
                        shape_str = self._format_shape_string(shapes)
                        
                        results.append({
                            'shape': shape_str,
                            'device': device_str,
                            'dtype': dtype,
                            'time_us': avg_time_us
                        })
                        
                    except Exception as e:
                        if verbose:
                            shape_str = self._format_shape_string(shapes)
                            print(f"Error testing {shape_str} {device_str} {dtype}: {e}", file=sys.stderr)
                        # 跳过失败的测试
        
        return results
    
    def run_from_command_line(self, argv=None) -> int:
        """从命令行运行基准测试（类似C++的run(int argc, char *argv[])）
        
        解析命令行参数，运行测试，输出结果
        
        Args:
            argv: 命令行参数列表，如果为None则使用sys.argv
            
        Returns:
            程序退出码（0表示成功）
        """
        import argparse
        
        parser = argparse.ArgumentParser(description=f'{self.get_operator_name()} operator PyTorch benchmark')
        parser.add_argument('-d', '--device', help='Device filter (cpu or cuda)')
        parser.add_argument('-s', '--shape', help='Tensor shape filter')
        parser.add_argument('-w', '--warmup', type=int, help='Number of warmup iterations (default: 5)')
        parser.add_argument('-r', '--repeat', type=int, help='Number of repeat iterations (default: 100)')
        parser.add_argument('-v', '--verbose', action='store_true', help='Verbose output')
        
        args = parser.parse_args(argv)
        
        # 运行测试
        results = self.run(
            device_filter=args.device,
            shape_filter=args.shape,
            warmup_cnt=args.warmup,
            repeat_cnt=args.repeat,
            verbose=args.verbose
        )
        
        # 输出结果（制表符分隔，与C++格式一致）
        if results:
            # 计算列宽
            max_shape_width = max(len('shape'), max(len(r['shape']) for r in results))
            repeat_str = str(args.repeat or 100)
            max_repeat_width = max(len('repeat'), len(repeat_str))
            max_device_width = max(len('device'), max(len(r['device']) for r in results))
            max_dtype_width = max(len('dtype'), max(len(r['dtype']) for r in results))
            max_time_width = max(len('pytorch_time_us'), max(len(f"{r['time_us']:.4f}") for r in results))
            
            # 输出表头
            print(f"{'shape':<{max_shape_width}}\t{'repeat':<{max_repeat_width}}\t{'device':<{max_device_width}}\t"
                  f"{'dtype':<{max_dtype_width}}\t{'pytorch_time_us':<{max_time_width}}")
            
            # 输出结果
            for r in results:
                print(f"{r['shape']:<{max_shape_width}}\t{repeat_str:<{max_repeat_width}}\t"
                      f"{r['device']:<{max_device_width}}\t{r['dtype']:<{max_dtype_width}}\t"
                      f"{r['time_us']:.4f}")
        else:
            print("No results generated", file=sys.stderr)
            return 1
        
        return 0
